crates/cranelift/src/compiler/component.rs

```diff
@@ -262,7 +258,163 @@ impl<'a> TrampolineCompiler<'a> {
+                            .map(|v| v.as_u32())
+                    })
+                    .unwrap_or(u32::MAX),
+            ),
+        );
+
+        let call = self.compiler.call_indirect_host(
+            &mut self.builder,
+            index,
+            host_sig,
+            host_fn,
+            &[vmctx, ty, values_vec_ptr, values_vec_len],
+        );
+        let succeeded = self.builder.func.dfg.inst_results(call)[0];
+        self.raise_if_host_trapped(succeeded);
+        self.builder.ins().return_(&[]);
+    }
+
+    fn translate_async_enter_or_exit(
+        &mut self,
+        get_libcall: fn(
+            &dyn TargetIsa,
+            &mut ir::Function,
+        ) -> (ir::SigRef, ComponentBuiltinFunctionIndex),
+        callback_and_post_return: Option<(
+            Option<RuntimeCallbackIndex>,
+            Option<RuntimePostReturnIndex>,
+        )>,
+        result: ir::types::Type,
+    ) {
+        match self.abi {
+            Abi::Wasm => {}
+
+            // These trampolines can only actually be called by Wasm, so
+            // let's assert that here.
+            Abi::Array => {
+                self.builder.ins().trap(TRAP_INTERNAL_ASSERT);
+                return;
+            }
+        }
+
+        let args = self.builder.func.dfg.block_params(self.block0).to_vec();
+        let vmctx = args[0];
+
+        let (host_sig, index) = get_libcall(self.isa, &mut self.builder.func);
+        let host_fn = self.load_libcall(vmctx, index);
+
+        let mut callee_args = vec![vmctx];
+
+        if let Some((callback, post_return)) = callback_and_post_return {
+            let pointer_type = self.isa.pointer_type();
+
+            // callback: *mut VMFuncRef
+            if let Some(callback) = callback {
+                callee_args.push(self.builder.ins().load(
+                    pointer_type,
+                    MemFlags::trusted(),
+                    vmctx,
+                    i32::try_from(self.offsets.runtime_callback(callback)).unwrap(),
+                ));
+            } else {
+                callee_args.push(self.builder.ins().iconst(pointer_type, 0));
+            }
+
+            // post_return: *mut VMFuncRef
+            if let Some(post_return) = post_return {
+                callee_args.push(self.builder.ins().load(
+                    pointer_type,
+                    MemFlags::trusted(),
+                    vmctx,
+                    i32::try_from(self.offsets.runtime_post_return(post_return)).unwrap(),
+                ));
+            } else {
+                callee_args.push(self.builder.ins().iconst(pointer_type, 0));
+            }
+        }
+
+        // remaining parameters
+        callee_args.extend(args[2..].iter().copied());
+
+        let call = self.compiler.call_indirect_host(
+            &mut self.builder,
+            index,
+            host_sig,
+            host_fn,
+            &callee_args,
+        );
```

Comment:
Like above, the `call_libcall` helper might be able to help clean this up a little further

Reply:
Per my response above: I've already done this in a follow-up PR.